### PR TITLE
fixed SDK issue while uploading app in debug scheme

### PR DESCRIPTION
**Pull request by arinjay**
## Summary

Problem - Error when trying to publish to Apple Store in debug scheme

Error thread - https://github.com/facebook/react-native/issues/31507

## Changelog

[iOS][Changed] - The diffs renames the required variable which was causing conflicts in names with Apple core SDK's

### CODE REVIEW

**Comment by arinjay** [comment]
@lunaleaps can you help with this PR,
one of the cases was failing earlier (timeout while running project) but my diffs don’t match the failure reason.

I need help in re-running the test workflows

**Comment by kelset** [comment]
@cortinico & @fortmarek this PR addressed the fact that if you try to upload to Apple's store a debug build, there seems to be an overlap of methods' names with the ones from the core iOS SDK. By changing them, it seems to address the problem - so I think overall it'd be a good change to integrate.
My only concern is that we might have to consider it a breaking change? I don't think we need to merge this for 0.69, but the issue has been open since May 2021 so it'd be good to get this sorted at some point.

(cc @tido64, do you have any opinions on this?)

**Comment by fortmarek** [comment]
> My only concern is that we might have to consider it a breaking change?

The properties changed are with `_` prefix and thus should be considered private. The only change which could be imho breaking is renaming of `handleNotification` to `handlePackageNotification`. Are we sure this is something we consider part of the public API? If so, it should be a breaking change, indeed.

I don't think there's a particular reason not to merge this change for 0.69.

File: React/Base/RCTKeyCommands.m [hunk 1]
```diff
@@ -18,10 +18,10 @@
 
 @interface UIEvent (UIPhysicalKeyboardEvent)
 
-@property (nonatomic) NSString *_modifiedInput;
-@property (nonatomic) NSString *_unmodifiedInput;
-@property (nonatomic) UIKeyModifierFlags _modifierFlags;
-@property (nonatomic) BOOL _isKeyDown;
+@property (nonatomic) NSString *_modifiedEventInput;
+@property (nonatomic) NSString *_unmodifiedInput;
```
Comment (tido64) [inline React/Base/RCTKeyCommands.m:22]: Nit: Trailing whitespace:
```suggestion
@property (nonatomic) NSString *_unmodifiedInput;
```

File: React/DevSupport/RCTPackagerClient.h [hunk 2]
```diff
@@ -25,7 +25,7 @@ extern const int RCT_PACKAGER_CLIENT_PROTOCOL_VERSION;
 @protocol RCTPackagerClientMethod <NSObject>
 
 - (void)handleRequest:(NSDictionary<NSString *, id> *)params withResponder:(RCTPackagerClientResponder *)responder;
-- (void)handleNotification:(NSDictionary<NSString *, id> *)params;
+- (void)handlePackageNotification:(NSDictionary<NSString *, id> *)params;
```
Comment (tido64) [inline React/DevSupport/RCTPackagerClient.h:28]: Shouldn't this be:

```suggestion
- (void)handlePackagerNotification:(NSDictionary<NSString *, id> *)params;
```

Is this change necessary to fix the uploading issue? It doesn't look like it would clash with anything. Apple's naming convention is usually something like `clientDidReceiveNotification:`. Since it's in a header file, it's kind of public.

File: React/DevSupport/RCTPackagerConnection.mm [hunk 3]
```diff
@@ -210,7 +210,7 @@ - (void)addHandler:(id<RCTPackagerClientMethod>)handler forMethod:(NSString *)me
 
   [self
       addNotificationHandler:^(NSDictionary<NSString *, id> *notification) {
-        [handler handleNotification:notification];
+        [handler handlePackageNotification:notification];
```
Comment (tido64) [inline React/DevSupport/RCTPackagerConnection.mm:213]: Same here:
```suggestion
        [handler handlePackagerNotification:notification];
```

File: React/DevSupport/RCTPackagerClient.h (hunk 2 above)
Comment (fortmarek) [reply React/DevSupport/RCTPackagerClient.h:28]: This is mentioned in the [original](https://github.com/facebook/react-native/issues/31507#issue-885801684) issue and is a part of the error message: 
> ITMS-90338: Non-public API usage - The app references non-public selectors in app: _isKeyDown, _modifiedInput, _modifierFlags, handleNotification:, isPassthrough, onSuccess:, removeValuesForKeys:completion:, setCategoryID:. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/

So it probably is a problem?

**Comment by cortinico** [comment]
Looping @cipolleschi in here.

> My only concern is that we might have to consider it a breaking change? I don't think we need to merge this for 0.69, but the issue has been open since May 2021 so it'd be good to get this sorted at some point.

IMHO we're fine merging this on `main` + mark it as a breaking change in the README. 

> I don't think there's a particular reason not to merge this change for 0.69.

Given the limited use case ("trying to publish to Apple Store in debug scheme") this is not pick-worthy for .69 imho.

**Comment by choubeyaakash77** [comment]
> Given the limited use case ("trying to publish to Apple Store in debug scheme") this is not pick-worthy for .69 imho.

Just curious about the impact of this breaking change here. From what I understand, since iOS doesn't allow users to install apps as Android does with apks, Testflight is the only available option to upload an iOS build and test for non-ios developers.

**Comment by choubeyaakash77** [comment]
>  Waiting 1 more release won't hurt more than how it is already doing.

@cipolleschi  Yeah it makes sense. If it is not a breaking change, there should be no problem in including it in 0.69, but if there is a breaking change, then that's a call on the react-native team I guess.

> submit something in Debug mode nor whether it is a good practice

Submitting on TestFlight is different from submitting on AppStore. I am not sure but I don't think submitting something in Debug mode is a bad practice. That's a limitation given to us by Apple.
There is a valid use case imo to float around an app of a different variant than Release, to some internal users. There are many reasons for doing that - an app for backend developers, adding specific options for variants that won't be available in a release app, etc. One can potentially achieve the same by maintaining different branches/tags but that brings an overhead of maintaining multiple codebases. And this is fairly common in Android to make things available by app's variant.

**Comment by cipolleschi** [comment]
Yeah, I see all the points and I agree with them.

What I was trying to say, failing, is that, if I remember well:
- submitting an app to the App Store is one thing, and IIRC it must be built for release.
- submitting an app for TestFlight should follow a different path and it can be sent in Debug mode.

So these warnings looks a little bit weird to me.

But again, it has a been a while since I had to submit something to Apple, and I was using an Enterprise profile to distribute the app internally in other context. So, I could remember wrong.

**Comment by arinjay** [comment]
@cortinico I can't see how it's a breaking change?

As per my understanding of the code, `RCTPackagerClient` is supposed to be an internal protocol used by SDK itself and not directly by the client.
So updating the method definition both at the protocol and implementation level can't break it on the client-side.

**Comment by arinjay** [comment]
@cipolleschi yes you’re right

1. submitting an app to the App Store (release mode) in one thing 
2. submitting an app for TestFlight(debug mode) is separate.

Schemes are different on iOS apps.

The need comes from the very limitation Apple has in the first place i.e. not being able to install IPA directly onto any iOS device.
This generally hampers the development cycle and the majority of folks have to resolve to release apps in debug mode on a test flight as not all end users have machines with Xcode.

Sharing it with multiple end-users and then gathering bugs becomes very easier. The whole iOS development cycle becomes more efficient.

**Comment by Tak-As** [comment]
When will this PR be merged?
I am facing this issue and am waiting for a merge.
I appreciate your help with moving this issue ahead.

**Comment by cortinico** [comment]
> When will this PR be merged?
> I am facing this issue and am waiting for a merge.
> I appreciate your help with moving this issue ahead.

Just to clarify, this won't be included in RN 0.69. The earliest this will be released in RN 0.70 as it's not considered a regression.

**Comment by gabriellend** [comment]
> Given the limited use case ("trying to publish to Apple Store in debug scheme") this is not pick-worthy for .69 imho.

Hey! My team is currently deadlocked until we find a way around this issue. Is this really a limited use case? We need to be able to upload to TestFlight in Debug mode in order to test before we release. Don't all ios developers need to do that? Maybe I'm missing something but that seems like a big problem. Thank you!

**Comment by javache** [comment]
@cipolleschi and I look at an alternative way to address the issue in https://github.com/facebook/react-native/issues/31507. For now this breaks some development workflows, so we prefer to back this out.

**Comment by arinjay** [comment]
@javache what's the alternate way of resolving this? Can you mention that

Also "_You should never upload debug scheme builds to the App Store_"
This is correct 
but we didn't upload it to AppStore, it failed to upload on the testFlight (these 2 are different mediums to publish app), even without the above MR changes.

Can you mention the changes using which we can still upload the app on testflight in debug mode?

**Comment by arinjay** [comment]
'_You should never upload debug scheme builds to any distribution platform_'
Why is that?
Apple made testfight for this purpose only. People to whom the app is made available are aware of the access they will be sharing when downloading the app from testflight. Apple made sure to list out all potential risks before setting up test flight apps.



**Use case in which debug mode is required** - An app running both on react native and native (swift) codebase. now in debug mode we need to have one API call debugger library or profiler library which we only want in debug mode only, and installed dependencies in debug mode only. Post putting it on test flight to share it with other user, which in real-time can track the profiler data on actual devices or detect which API was fired.

These cases are very much required in many day-to-day app developments.

**Comment by gabriellend** [comment]
Not sure this is still an ongoing conversation as the last reply was in June but I don't think it was implemented as I don't see it in the tags. I wanted to throw my two cents in as I can attest to the importance of being able to upload a debug scheme to TestFlight for internal testing before we submit a release version to the App Store. We've been doing it this way for years at my job and we now can't submit to TestFlight because of running into the selectors mentioned in this post. I've been in talks with Apple and they just got back to me directing me to manually change these myself so that the upload will be accepted to TestFlight. I want to voice my support for including this in the package. Thanks!

**Comment by rolfen** [comment]
@javache  Whether you should upload debug scheme builds to the App Store or not is out or RN's scope.

This error message seems to suggest that something wrong is happening.
